src/websocket_server/websocket_server.c

```diff
@@ -0,0 +1,157 @@
+/**
+ * @file websocket_server.c  Weboscket Server Demo
+ *
+ * The demo can for instance be tested with the telnet tool:
+ *  var ws = new WebSocket('ws://localhost:3456');
+ *  ws.addEventListener('open', () => ws.send('hello'));
+ *  ws.addEventListener('close', (e) => console.log('close', e));
+ *  ws.addEventListener('message', (e) => console.log('msg', e));
+ *
+ * Copyright (C) 2018 Creytiv.com
+ */
+
+#include <string.h>
+#include <re.h>
+
+
+struct websock *server;
+struct websocket {
+    struct le le;               /* list entry */
+    struct websock_conn *conn;  /* Websocket connection */
+};
+
+/* Linked list of active Websocket connections */
+static struct list connl;
+
+/* called upon reception of  SIGINT, SIGALRM or SIGTERM */
+static void signal_handler(int sig)
+{
+    re_printf("terminating on signal %d...\n", sig);
+
+    /* destroy active Websocket connections */
+    list_flush(&connl);
+
+    websock_shutdown(server);
+}
+
+static void websock_shutdown_handler(void *arg)
+{
+    (void)arg;
+    re_printf("websocket server shutting down\n");
+
+    /* stop libre main loop */
+    re_cancel();
+}
+
+/* destructor called when reference count on conn reach zero */
+static void destructor(void *arg)
+{
+    struct websocket *ws = arg;
+
+    /* remove this connection entry from the list */
+    list_unlink(&ws->le);
+
+    /* clean up local Websocket connection state */
+    mem_deref(ws->conn);
+}
+
+/* called when a message is received from the other side */
+static void srv_websock_recv_handler(const struct websock_hdr *hdr,
+                     struct mbuf *mb, void *arg)
+{
+    struct websocket *ws = arg;
+    int err;
+
+    err = websock_send(ws->conn, hdr->opcode, "%b", mbuf_buf(mb), mbuf_get_left(mb));
+    if (err)
+        re_printf("ws send error: %s\n", strerror(err));
```

Comment:
instead of strerror, it is now better to use the re-specific %m formatting:

```c
re_printf("error: %m\n", err);
```

(this goes for all entries in the file)